mmcv/ops/nms.py

```diff
@@ -141,7 +161,10 @@ def nms(boxes, scores, iou_threshold, offset=0):
         }
         inds = ext_module.nms(*indata_list, **indata_dict)
     else:
-        inds = NMSop.apply(boxes, scores, iou_threshold, offset)
+        if max_num < 0:
```

Comment:
we could pass `max_num` to nms even though max_num <= 0.

Reply:
If we pass `max_num < 0` to the `NMSop.forward` function, then during export this value will be written to the ONNX graph.
This is not good, so we need to prepare the correct arguments before calling this function.

Reply:
if `max_num < 0` , then we could set `max_output_per_class` to `sys.maxsize` as original style:
```
            max_output_per_class = g.op(
                'Constant',
                value_t=torch.tensor([sys.maxsize], dtype=torch.long))
```

Reply:
It seems to me that using `sys.maxsize` significantly overestimates the value of the `max_output_per_class` argument.

The number of indices after the NMS operation can be limited by the size of `bboxes`.
Then, if `max_num` is not specified or `max_num <= 0`, we can calculate its value in the ONNX graph using `bbox` shape.

Reply:
Agree, `sys.maxsize` is indeed too large.

Reply:
However, there's another issue. `max_output_per_class` should be constant for TensorRT backend. Using `bbox.shape` as `max_num` would create dynamic `max_output_per_class`.

Reply:
So, if this limitation only applies to TensorRT, then perhaps it should only be considered when exporting to TensorRT?
For example, we can replace `bbox.shape` with a constant value somewhere in this function:
https://github.com/open-mmlab/mmcv/blob/5d75c7a8ee4a0710208a86c75bc654551023b0f3/mmcv/tensorrt/tensorrt_utils.py#L7

Reply:
Normally,  `max_num` should be required to set when exporting to ONNX for deployment.

Reply:
If there is one NMS operation in the model, then the `max_num` value in the config will be specified for it.
But for example in Faster RCNN, there are two NMS operations.
For the last operation (from BBoxHead) the `max_num` value is specified as a constant.
For NMS operation from RPNHead, when executed in PyTorch, the `max_num` value is not specified, that is, all indexes are used in further processing.
It would be correct to keep this behavior in ONNX as well.

Dynamically defining `max_num` through the `bbox.shape` in the ONNX graph allows us to use the reshaping of the model.
It is possible that when the input resolution is changed, the stored constant value of `max_num` in NMS from RPNHead will be incorrect.

Reply:
`max_num` for RPNHead is set in the [config](https://github.com/open-mmlab/mmdetection/blob/2f2cb37cdaeac0862710045a066ee1669e286a42/configs/_base_/models/faster_rcnn_r50_fpn.py#L99).

Reply:
Yes, I didn't notice that filtering by `max_per_img` is applied after calling `batched_nms`. Then nms was called without limitation on `max_output_boxes_per_class` and the `sys.maxsize` value was stored in the ONNX graph.

I will add the `max_per_img` value to `cfg.nms`.